File: Java/Tema8_Ejercicio8.java
/*
 * TEMA 8: ESTRUCTURAS DE DATOS AVANZADAS - EJERCICIO 8
 * ======================================================
 * Escriba un programa que implemente una lista de tareas con prioridades.
 */

import java.util.HashMap;
import java.util.ArrayList;
import java.util.Map;

public class Tema8_Ejercicio8 {
    
    public static void main(String[] args) {
        // HashMap para guardar tareas por prioridad
        HashMap<String, ArrayList<String>> tareas = new HashMap<>();
        
        // Crear listas para cada prioridad
        tareas.put("Alta", new ArrayList<>());
        tareas.put("Media", new ArrayList<>());
        tareas.put("Baja", new ArrayList<>());
        
        // Agregar tareas
        tareas.get("Alta").add("Finalizar proyecto");
        tareas.get("Alta").add("Entregar reportes");
        
        tareas.get("Media").add("Responder emails");
        tareas.get("Media").add("Actualizar documentación");
        tareas.get("Media").add("Revisar código");
        
        tareas.get("Baja").add("Organizar escritorio");
        tareas.get("Baja").add("Limpiar archivos");
        
        System.out.println("=== LISTA DE TAREAS POR PRIORIDAD ===");
        
        // Mostrar tareas
        String[] prioridades = {"Alta", "Media", "Baja"};
        for (String prioridad : prioridades) {
            ArrayList<String> listaTareas = tareas.get(prioridad);
            System.out.println("\n--- Prioridad " + prioridad + " (" + 
                             listaTareas.size() + " tareas) ---");
            
            for (int i = 0; i < listaTareas.size(); i++) {
                System.out.println("  " + (i + 1) + ". " + listaTareas.get(i));
            }
        }
        
        // Contar total de tareas
        int totalTareas = 0;
        for (ArrayList<String> lista : tareas.values()) {
            totalTareas += lista.size();
        }
        System.out.println("\nTotal de tareas: " + totalTareas);
        
        // Marcar tarea como completada
        System.out.println("\n--- Marcando 'Finalizar proyecto' como completada ---");
        tareas.get("Alta").remove("Finalizar proyecto");
        System.out.println("Tareas Alta restantes: " + tareas.get("Alta"));
    }
}

/*
 * SALIDA ESPERADA:
 * === LISTA DE TAREAS POR PRIORIDAD ===
 * 
 * --- Prioridad Alta (2 tareas) ---
 *   1. Finalizar proyecto
 *   2. Entregar reportes
 * 
 * --- Prioridad Media (3 tareas) ---
 *   1. Responder emails
 *   2. Actualizar documentación
 *   3. Revisar código
 * 
 * --- Prioridad Baja (2 tareas) ---
 *   1. Organizar escritorio
 *   2. Limpiar archivos
 * 
 * Total de tareas: 7
 * 
 * --- Marcando 'Finalizar proyecto' como completada ---
 * Tareas Alta restantes: [Entregar reportes]
 */

File: JS/Tema8_Ejercicio8.js
/*
 * TEMA 8: ESTRUCTURAS DE DATOS AVANZADAS - EJERCICIO 8
 * =====================================================
 * Sistema de manejo de tareas por prioridad.
 * 
 * EXPLICACIÓN:
 * - Organizar tareas por nivel de prioridad
 * - Usar Map para prioridades
 * - Operaciones: agregar, completar, listar
 */

class ManejadorTareas {
    constructor() {
        this.tareas = new Map([
            ["ALTA", []],
            ["MEDIA", []],
            ["BAJA", []]
        ]);
    }
    
    agregarTarea(descripcion, prioridad = "MEDIA") {
        if (this.tareas.has(prioridad)) {
            this.tareas.get(prioridad).push({
                descripcion,
                completada: false
            });
        }
    }
    
    completarTarea(prioridad, indice) {
        const tareas = this.tareas.get(prioridad);
        if (tareas && tareas[indice]) {
            tareas[indice].completada = true;
        }
    }
    
    listarTareas() {
        console.log("=== GESTOR DE TAREAS ===\n");
        let numeroTotal = 0;
        
        this.tareas.forEach((tareas, prioridad) => {
            if (tareas.length > 0) {
                console.log(`${prioridad}:`);
                tareas.forEach((t, i) => {
                    const estado = t.completada ? "✓" : "○";
                    console.log(`  ${estado} ${i + 1}. ${t.descripcion}`);
                    numeroTotal++;
                });
                console.log();
            }
        });
        
        console.log(`Total de tareas: ${numeroTotal}`);
    }
}

const gestor = new ManejadorTareas();

gestor.agregarTarea("Arreglar bug crítico", "ALTA");
gestor.agregarTarea("Llamar al cliente", "ALTA");
gestor.agregarTarea("Revisar código", "MEDIA");
gestor.agregarTarea("Escribir documentación", "MEDIA");
gestor.agregarTarea("Limpiar escritorio", "BAJA");

gestor.completarTarea("ALTA", 0);
gestor.completarTarea("MEDIA", 0);

gestor.listarTareas();

/*
 * SALIDA ESPERADA:
 * === GESTOR DE TAREAS ===
 * 
 * ALTA:
 *   ✓ 1. Arreglar bug crítico
 *   ○ 2. Llamar al cliente
 * 
 * MEDIA:
 *   ✓ 1. Revisar código
 *   ○ 2. Escribir documentación
 * 
 * BAJA:
 *   ○ 1. Limpiar escritorio
 * 
 * Total de tareas: 5
 */
